---
title: "DESeq analysis, Volcano plot and PCA plot"
author: "21267455 Ella Zhang"
date: "2024-11-25"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(DESeq2)
library(ggplot2)
library(pheatmap)
library(clusterProfiler)
library(org.Hs.eg.db)
library(data.table)
library(readr)
library(dplyr)
```

```{r}
clinical.patients <- read.table("data_clinical_patient.txt", sep = "\t", header = TRUE)
data.mutations <- read.table("data_mutations.txt", sep = "\t", header = TRUE)
data.RNAseq <- read.csv("RNAseq_BRCA.csv")
```

```{r, message=FALSE}
library(stringr)

# Rename columns to match the desired format
colnames(data.RNAseq) <- sapply(colnames(data.RNAseq), function(name) {
  
  segments <- strsplit(name, "\\.")[[1]][1:3]
  
  paste(segments, collapse = "-")
})

colnames(data.RNAseq)[1] <- "Transcript_ID"

```

```{r}
unique.clinical <- as.data.frame(unique(clinical.patients$PATIENT_ID))
unique.mutations <- as.data.frame(unique(data.mutations$Tumor_Sample_Barcode))
unique.RNA <- as.data.frame(colnames(data.RNAseq[,2:1232]))

colnames(unique.clinical) <- "Patient_ID"
colnames(unique.mutations) <- "Patient_ID"
colnames(unique.RNA) <- "Patient_ID"

unique.mutations$Patient_ID <- substr(unique.mutations$Patient_ID, 1, 12)

# Find common patient IDs across all three data frames
common_patient_ids <- Reduce(intersect, list(unique.clinical$Patient_ID, unique.mutations$Patient_ID, unique.RNA$Patient_ID))

filtered.clinical <- clinical.patients[clinical.patients$PATIENT_ID %in% common_patient_ids, ]
filtered.mutations <- data.mutations[substr(data.mutations$Tumor_Sample_Barcode, 1, 12) %in% common_patient_ids, ]
filtered.RNAseq <- data.RNAseq[, c("Transcript_ID", common_patient_ids)]  #Keep only the columns of common patients
```

```{r}
RNAseq_numeric <- as.matrix(filtered.RNAseq[, -1])
rownames(RNAseq_numeric) <- filtered.RNAseq$Transcript_ID

filtered.clinical$SurvivalStatus <- ifelse(filtered.clinical$OS_MONTHS > 36, "HighSurvival", "LowSurvival")
filtered.clinical$SurvivalStatus <- as.factor(filtered.clinical$SurvivalStatus)

dds <- DESeqDataSetFromMatrix(countData = RNAseq_numeric,
                              colData = filtered.clinical,
                              design = ~ SurvivalStatus)
```

```{r}
dds <- DESeq(dds)

res <- results(dds, contrast = c("SurvivalStatus", "HighSurvival", "LowSurvival"))

res <- lfcShrink(dds, coef = 2, type = "apeglm")

summary(res)

sig_res <- res[which(res$padj < 0.05), ]

write.csv(as.data.frame(sig_res), "DEGs_results.csv")
```

There were 57,944 genes included in the analysis, genes with zero counts in all samples were excluded. the p value is less than 0.1, which means this data is considered statistical significant.

```{r}
res_df <- as.data.frame(res)

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "red") +
  xlab("Log2 Fold Change") +
  ylab("-Log10 Adjusted P-value") +
  ggtitle("Volcano Plot of Differential Expression")
```

This plot identifies a small subset of genes with significant differential expression, their biological roles can be investigated in survival outcomes

```{r}
# Perform PCA
vsd <- vst(dds, blind = FALSE)
pca_data <- plotPCA(vsd, intgroup = "SurvivalStatus", returnData = TRUE)

# Visualize PCA
ggplot(pca_data, aes(PC1, PC2, color = SurvivalStatus)) +
  geom_point(size = 3) +
  ggtitle("PCA Plot of RNA-Seq Data") +
  xlab("PC1") +
  ylab("PC2")
```

This plot shows the partial separation between survival groups, suggesting that survival status can be one of the factors that can influence the gene expression.
